### PR TITLE
Add undo-redo UI to edit-site and edit-widgets

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR adds the Undo redo UI to the edit-site and edit-widgets screens, making them consistent with edit-post.

The undo-redo does not work well on these screens yet. Some problems are:
- On the widget screen, when applying changes to multiple areas, sometimes the undo just undo the changes in one of the areas (I think it is a bug in core-data).
- On the widget scree on paragraphs, it seems the undo is character by character (I think we are probably mishandling changes in BlockEditorProvider inside widgets screen).
- On the edit-site screen, it appears the undo works in some cases but ignores changes in some template areas.

This PR just adds the UI, so the issues we have become noticeable. I will follow up with some bug fixes to the problems we currently have. I think exposing the UI even if undo-redo is not working well is a first step to fix all the issues.

## How has this been tested?
I verified the undo and redo buttons appeared on the edit-site and edit-widgets screens.

### CODE REVIEW

File: packages/edit-widgets/src/components/header/undo-redo/undo.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Button } from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { undo as undoIcon } from '@wordpress/icons';
+import { displayShortcut } from '@wordpress/keycodes';
+
+export default function UndoButton() {
```
Comment (noahtallen) [inline packages/edit-widgets/src/components/header/undo-redo/undo.js:10]: Is there a package these components could go in so we aren't duplicating code between edit post, edit site, and edit widgets?

**Comment by noahtallen** [comment]
> appears the undo works in some cases but ignores changes in some template areas.

I'm also testing undo in https://github.com/WordPress/gutenberg/pull/21368, but I've been trying it by inserting template parts in the post editor. My PR is mostly about the dirty state, but it changes how some of the blocks get in and out of the block editor package (and how they relate to their entities), so we have been looking for regressions. I'm "glad" so see that the behavior is a bit buggy despite my code 😁

File: packages/edit-widgets/src/components/header/undo-redo/undo.js (hunk 1 above)
Comment (vindl) [reply packages/edit-widgets/src/components/header/undo-redo/undo.js:10]: Maybe `block-editor` or alternatively `interface`?

Comment (youknowriad) [reply packages/edit-widgets/src/components/header/undo-redo/undo.js:10]: I wouldn't bother sharing these for the moment personally.

File: packages/edit-site/src/components/header/undo-redo/redo.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Button } from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { redo as redoIcon } from '@wordpress/icons';
+import { displayShortcut } from '@wordpress/keycodes';
+
+export default function UndoButton() {
+	const hasRedo = useSelect( ( select ) => select( 'core' ).hasUndo() );
```
Comment (Addison-Stavlo) [inline packages/edit-site/src/components/header/undo-redo/redo.js:11]:
```suggestion
	const hasRedo = useSelect( ( select ) => select( 'core' ).hasRedo() );
```

Comment (Addison-Stavlo) [inline packages/edit-site/src/components/header/undo-redo/redo.js:11]:
```suggestion
	const hasRedo = useSelect( ( select ) => select( 'core' ).hasRedo() );
```
I assume this should be using the `hasRedo` selector. 😄

File: packages/edit-widgets/src/components/header/undo-redo/redo.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Button } from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { redo as redoIcon } from '@wordpress/icons';
+import { displayShortcut } from '@wordpress/keycodes';
+
+export default function UndoButton() {
+	const hasRedo = useSelect( ( select ) => select( 'core' ).hasUndo() );
```
Comment (Addison-Stavlo) [inline packages/edit-widgets/src/components/header/undo-redo/redo.js:11]:
```suggestion
	const hasRedo = useSelect( ( select ) => select( 'core' ).hasRedo() );
```
Similarly here ☝️

**Review by Addison-Stavlo — COMMENTED** [review]
This is looking good so far.  🎉 

Just a minor comment/suggestion below, it looks like its using the hasUndo selector in the redo components.

**Comment by jorgefilipecosta** [comment]
Hi @noahtallen, @Addison-Stavlo thank you for the comments, review, and suggestions the changes were applied 👍

**Comment by noahtallen** [comment]
The undo buttons seem to be working ok in edit site, but for some reason cmd-z isn't undoing the changes. It's actually weird, so it might be on my side:

- I click cmd-z a ton of times in a row
- undo does not happen
- I open the "Edit" menu in the macOS toolbar and then do cmd-z again (or click the Undo menu item there)
- undo happens

It happened for me in brave browser and in firefox dev edition, so not sure if it's just my environment, or if it's something about the shortcuts

**Comment by Addison-Stavlo** [comment]
>The undo buttons seem to be working ok in edit site, but for some reason cmd-z isn't undoing the changes. It's actually weird, 

Confirming this as well.  The buttons work but the keyboard shortcuts don't seem to.  Perhaps this needs to be configured for edit-site somewhere? 🤔

**Review by noahtallen — APPROVED** [review]
It's working for me now, thanks! I still feel like it's weird to be duplicating nearly identical code across several packages, but I don't think it should be a blocker for starting to have this functionality and fixing the more important issues with undo & e.g. template parts :)

File: packages/edit-widgets/src/components/keyboard-shortcuts/index.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	);
+
+	return null;
+}
+function KeyboardShortcutsRegister() {
+	// Registering the shortcuts
+	const { registerShortcut } = useDispatch( 'core/keyboard-shortcuts' );
+	useEffect( () => {
+		registerShortcut( {
+			name: 'core/editor/undo',
```
Comment (Addison-Stavlo) [inline packages/edit-widgets/src/components/keyboard-shortcuts/index.js:37]: In edit-post it looks like we are using naming convention of `core/edit-post/undo`, should these be `core/edit-site/undo` / similarly for widgets?

**Review by Addison-Stavlo — APPROVED** [review]
Thanks for adding the keyboard shortcuts!  This seems to be working for me as well.

For the site editor I guess the best way to test undo/redo is with the site title, as the functionality with child entities is still a bit strange and being fleshed out (although thats a different issue)

In widgets, it seems the undo/redo only works for blocks that are added and not the forms themselves.  But I don't know what goals are there.

For the scope of this PR, this seems to be working well as far as I can tell.  Only a small nit on naming (if that happens to be any issue there).

File: packages/edit-widgets/src/components/keyboard-shortcuts/index.js (hunk 4 above)
Comment (jorgefilipecosta) [reply packages/edit-widgets/src/components/keyboard-shortcuts/index.js]: Hi @Addison-Stavlo nice catch, I updated the code 👍

**Pull request merged**